### PR TITLE
last version

### PR DESCRIPTION
**Pull request by xpazv**
Version with the latest changes

### CODE REVIEW

File: lang/en/local_aspiredu.php [hunk 1]
```diff
@@ -56,3 +56,6 @@
 $string['privacy:metadata:aspiredu_provider:lastname'] = 'The last name of the user.';
 $string['privacy:metadata:aspiredu_provider:userid'] = 'The ID of the user.';
 $string['privacy:metadata:aspiredu_provider:username'] = 'The username of the user.';
+// New for lti 1.3
+$string['instance'] = 'Tool instance id';
```
Comment (tim-schilling) [inline lang/en/local_aspiredu.php:67]: ~~What is a typical value for this @xpazv?~~

**Comment by tim-schilling** [comment]
~~@xpazv does this require moodle v4.0?~~

File: lang/en/local_aspiredu.php (hunk 1 above)
Comment (tim-schilling) [reply lang/en/local_aspiredu.php:67]: We need to support ~~four~~ two different types of product integrations.

- Dropout Detective
- ~~Grade Guardian~~
- Instructor Insight
- ~~Retention Radar~~

This user experience of this flow isn't ideal. If they need to configure the external tools elsewhere (http://134.209.160.50/moodlegit/admin/settings.php?section=modsettinglti), why are we needing them to select something here? It should be possible to look up that tool where this parameter is being used.

**Comment by tim-schilling** [comment]
@xpazv I have this somewhat working.

I am entirely new to LTI within Moodle, so I had to guess my way through all of this. I would appreciate any advice/corrections/guides.

1. I updated the external tool here: http://134.209.160.50/moodlegit/admin/settings.php?section=modsettinglti (how did this get setup in the first place?)
2. I collected the information on the tool from here: http://134.209.160.50/moodlegit/mod/lti/toolconfigure.php
3. I created the tool instance in our system here with the values from the above external tool: https://aspiredu-bld-3ipunt.herokuapp.com/admin/lti/tool/1/change/
4. I created the tool integration in our system connecting the tool to the deployment id: https://aspiredu-bld-3ipunt.herokuapp.com/admin/lti/toolintegration/1/change/
5. I finally had to create an activity with the external tool (http://134.209.160.50/moodlegit/course/modedit.php?update=69&return=0&sr=0) We don't want to require this step.

--

This gets the flow through the oidc initation, back to Moodle, back to our system for authentication, but the payload being sent to the server is malformed. It has an extra trailing closing curly bracket:

```pyterm
In [24]: decoded['payload'].decode('utf-8')
Out[24]: '{"nonce":"47c5f011b7284e73a8747ffd626dfdbf359d3dd871b111ed9252e62c885300aa","iat":1669924144,"exp":1669924204,"iss":"http:\\/\\/134.209.160.50\\/moodlegit","aud":"zzO0uPWEDeGuSOB","https:\\/\\/purl.imsglobal.org\\/spec\\/lti\\/claim\\/deployment_id":"1","https:\\/\\/purl.imsglobal.org\\/spec\\/lti\\/claim\\/target_link_uri":"https:\\/\\/aspihip.status","membership_role":"Instructor,urn:lti:sysrole:ims\\/lis\\/Administrator,urn:lti:instrole:ims\\/lis\\/Administrator"}}'
```

This is what it's trying to be:
```python
{'nonce': '47c5f011b7284e73a8747ffd626dfdbf359d3dd871b111ed9252e62c885300aa',
 'iat': 1669924144,
 'exp': 1669924204,
 'iss': 'http://134.209.160.50/moodlegit',
 'aud': 'zzO0uPWEDeGuSOB',
 'https://purl.imsglobal.org/spec/lti/claim/deployment_id': '1',
 'https://purl.imsglobal.org/spec/lti/claim/target_link_uri': 'https://aspihip.status',
 'membership_role': 'Instructor,urn:lti:sysrole:ims/lis/Administrator,urn:lti:instrole:ims/lis/Administrator'}
```

This results in an error on our server because it expects valid JSON, which the above is not due to the extra `}` at the end.

Additionally, the actual packet of data does not have the data we need. We're looking for the user and course ids which are not included.

--

The other thing is that the above scenario is only when launching through Moodle's own External Tool system. When trying to use the Report > Dropout Detective link from either the dashboard or a course, it results in the following error:

http://134.209.160.50/moodlegit/local/aspiredu/aspiredu.php?id=1&product=dd

```
Can't find data record in database.

[More information about this error](https://docs.moodle.org/310/en/error/moodle/invalidrecordunknown)
Debug info: SELECT cm.*, m.name, md.name AS modname
FROM {course_modules} cm
JOIN {modules} md ON md.id = cm.module
JOIN {lti} m ON m.id = cm.instance

WHERE cm.id = :cmid AND md.name = :modulename

[array (
'cmid' => '1',
'modulename' => 'lti',
)]
Error code: invalidrecordunknown
Stack trace:

    line 1674 of /lib/dml/moodle_database.php: dml_missing_record_exception thrown
    line 1246 of /lib/datalib.php: call to moodle_database->get_record_sql()
    line 116 of /mod/lti/auth.php: call to get_coursemodule_from_id()
```


http://134.209.160.50/moodlegit/local/aspiredu/aspiredu.php?id=5&product=dd
```
Can't find data record in database.

[More information about this error](https://docs.moodle.org/310/en/error/moodle/invalidrecordunknown)
Debug info: SELECT cm.*, m.name, md.name AS modname
FROM {course_modules} cm
JOIN {modules} md ON md.id = cm.module
JOIN {lti} m ON m.id = cm.instance

WHERE cm.id = :cmid AND md.name = :modulename

[array (
'cmid' => '5',
'modulename' => 'lti',
)]
Error code: invalidrecordunknown
Stack trace:

    line 1674 of /lib/dml/moodle_database.php: dml_missing_record_exception thrown
    line 1246 of /lib/datalib.php: call to moodle_database->get_record_sql()
    line 116 of /mod/lti/auth.php: call to get_coursemodule_from_id()
```

File: lang/en/local_aspiredu.php (hunk 1 above)
Comment (xpazv) [reply lang/en/local_aspiredu.php:67]: As it is implemented, right now, the user has to set up the external tool, at site level, and then, will have a block, available in dashboard or course context, where we show the user the access to the product integrations (currently DD and ii)

**Comment by xpazv** [comment]
Regarding the first part of the comment, the integration, with the latest commits, we were already able to connect DD and ii systems, according to the requirements agreed in the meetings.

File: lang/en/local_aspiredu.php (hunk 1 above)
Comment (tim-schilling) [reply lang/en/local_aspiredu.php:67]: I check with Kim. We're sticking with only DD and II. The other products are being tabled for now.

File: db/access.php [hunk 2]
```diff
@@ -40,6 +41,7 @@
         'contextlevel' => CONTEXT_COURSE,
         'archetypes' => array(
             'editingteacher' => CAP_ALLOW,
+            'student' => CAP_ALLOW,
```
Comment (tim-schilling) [inline db/access.php]: @xpazv this comment seems to have been lost on that hanging commit:

> Student's are never allowed to access any of our products. Does that change this?

Comment (xpazv) [reply db/access.php]: Wi will set this capability to prohibit for student role

Comment (tim-schilling) [reply db/access.php]: Does that mean that this line needs to change?

Comment (xpazv) [reply db/access.php]: Yes, it will be changed in the next commit from our side.

**Pull request closed without merging**